orttraining/orttraining/python/training/ortmodule/_inference_manager.py

```diff
@@ -73,27 +66,38 @@ def forward(self, *inputs, **kwargs):
             return self._fallback_manager.fallback(self._original_module, self._debug_options.logging.log_level, *inputs, **kwargs)
 
         try:
-            # Exporting module to ONNX for the first time
-            build_graph = self._export_model(*inputs, **kwargs)
-            if build_graph:
-                # If model was exported, then initialize the graph builder
-                self._initialize_graph_builder(training=False)
-
-            # Build the inference graph
-            if build_graph:
-                self._build_graph()
-
-            module_device = _utils.get_device_from_module(
-                self._original_module)
-            # The inference session should be created every time
-            # the graph was built or if the device changed between calls to forward
-            create_execution_session = build_graph or self._device != module_device
-            if self._device != module_device:
-                self._device = module_device
+            build_graph = False
+            if self._skip_check.is_set(_SkipCheck.SKIP_CHECK_BUILD_GRADIENT) == False or \
```

Comment:
Maybe adding a `warns.warn` for each time "skip check" actually skips something? It certainly helps debugging, to distinguish a skip from an actual bug

Reply:
I think adding a warning for every forward might add unnecessary printing. When loading the json config, we do log a statement letting the user's know that which config was loaded. I think that should suffice.

Reply:
Agreed, repeated logs during training for the same execution path is not ideal. A single warning (for both regular and json API users) should suffice. Logs for load_from_json is not an ideal UX, as it doesn't provide debug info for the mainstream users, which dont depend on json

```
if self._debug_options.logging.log_level <= _logger.LogLevel.WARNING:
    warnings.warn("One or more model checks were skipped for faster execution.", UserWarning)
```

Reply:
This get's called repeatedly on every forward, so putting that debug line here will result in many repeated log lines.
Adding to the ORTModule constructor does not make sense either right now, because the loading from json happens after the constructor has completed its execution.
Which is why we only have 1 log statement in the load_from_json file.

Reply:
(for some reason, github is not allowing inline comments today)

You are right, the current implementation doesn't help achieving unique logs right out of the bat, but this PR removes an existing debug log for informing that fast path is enabled. It is repeated, but it is logged only when log level is warning (not the default) and when fast path is enabled (not the default).

We should either keep the dumb log as it is or improve it, which would be my preference

One idea is to add a bool `self._first_forward_call=True` (during init) that flips to False at the end of `forward` so that any piece of code that might need to log something on the first forward call could use it. The fact `self._first_forward_call` is True even before the first `forward` is executed shouldn't be a problem because this flag is meant to be used by `forward` only. A comment with that intent might be added to clarify

Kshama and Priscilla asked about debugability in the last meeting, which are important for devs that are onboarding models into ORTModule. That might be an useful/lightweight improvement for informing the current state of skip check and possibly other features that do not provide insightful logs for both internal and public settings.